### PR TITLE
[2198] Do not blindly set #threads to requested #threads (4-3-stable)

### PR DESCRIPTION
**Pull request by JustinKyleJames**
This is the same as pull request #2199 but with Python3 changes as well as a slight change in output when iput -r is run in 4.3.

### CODE REVIEW

File: packaging/resource_suite_s3_nocache.py [hunk 1]
```diff
@@ -2406,6 +2409,67 @@ def test_upload_very_large_file_after_redirect_issues_2104_2114(self):
...
+                assert_command(['diff', '-q', filename, get_filename])
+
+                for n_value in [0, 1, 2,3,4,8]:
+                    self.user1.assert_icommand(f'iput -f -N{n_value} {filename}')
+                    self.user1.assert_icommand(f'iget -f {filename} {get_filename}')
+                    assert_command(['diff', '-q', filename, get_filename])
+
+                # cleanup before next loop
+                s3plugin_lib.remove_if_exists(filename)
+                s3plugin_lib.remove_if_exists(get_filename)
```
Comment (alanking) [inline packaging/resource_suite_s3_nocache.py:2441]: Consider including these two lines in the `finally` block as well, in the event of a failure in the loop above.

File: packaging/resource_suite_s3_nocache.py [hunk 2]
```diff
@@ -2406,6 +2409,67 @@ def test_upload_very_large_file_after_redirect_issues_2104_2114(self):
...
+            for file_size in file_sizes:
+                # create the file
+                lib.make_arbitrary_file(filename, file_size)
+
+                # put the file
+                self.user1.assert_icommand(f'iput {filename}')
+                self.user1.assert_icommand(f'iget -f {filename} {get_filename}')
+                assert_command(['diff', '-q', filename, get_filename])
+
+                for n_value in [0, 1, 2,3,4,8]:
```
Comment (alanking) [inline packaging/resource_suite_s3_nocache.py:2434]: Let's space these out consistently.

File: packaging/resource_suite_s3_nocache.py [hunk 3]
```diff
@@ -2406,6 +2409,67 @@ def test_upload_very_large_file_after_redirect_issues_2104_2114(self):
...
+            os.mkdir(dirname)
+            # create the files
+            lib.make_arbitrary_file(f'{dirname}/{filename1}', 100)
+            lib.make_arbitrary_file(f'{dirname}/{filename2}', 100)
+            lib.make_arbitrary_file(f'{dirname}/{filename3}', 100)
+
+            # recursively put the directory
+            self.user1.assert_icommand(f'iput -r -N2 -K {dirname}')
+
+            #verify the files are in iRODS
```
Comment (alanking) [inline packaging/resource_suite_s3_nocache.py:2465]: Should we also verify the contents in some way? Don't necessarily need to verify checksums, but maybe an `iget` with a `diff` like the other test as well as maybe checking for the existence of a checksum in the catalog could be enough... Not sure if that's overkill for what this is testing. If so, maybe an explanatory comment or more words in the test name could explain that?

**Review by alanking — COMMENTED** [review]
Seems fine. Had some suggestions for the test.

File: packaging/resource_suite_s3_nocache.py (hunk 1 above)
Comment (JustinKyleJames) [reply packaging/resource_suite_s3_nocache.py:2441]: done

File: packaging/resource_suite_s3_nocache.py (hunk 2 above)
Comment (JustinKyleJames) [reply packaging/resource_suite_s3_nocache.py:2434]: done

File: packaging/resource_suite_s3_nocache.py (hunk 3 above)
Comment (JustinKyleJames) [reply packaging/resource_suite_s3_nocache.py:2465]: done

**Comment by JustinKyleJames** [comment]
I made the suggested changes.  I also appended changes for #2204 onto this PR.

File: s3_resource/src/s3_operations.cpp [hunk 4]
```diff
@@ -327,13 +327,11 @@ namespace irods_s3 {
...
             }
 
-            // if number of threads was not successfully set above
+            // If number of threads was not successfully set above.
             if (number_of_threads == 0) {
-
                 const int single_buff_sz = irods::get_advanced_setting<const int>(irods::KW_CFG_MAX_SIZE_FOR_SINGLE_BUFFER) * 1024 * 1024;
-                number_of_threads = requested_number_of_threads;
 
-                if (data_size > single_buff_sz && oprType != REPLICATE_DEST && oprType != COPY_DEST) {
```
Comment (korydraughn) [inline s3_resource/src/s3_operations.cpp:336]: Did we reach a conclusion about the `oprType` conditions?

File: packaging/resource_suite_s3_nocache.py [hunk 5]
```diff
@@ -2488,6 +2488,120 @@ def recursive_put_small_files_with_checksum_and_N_flag__issue_2198(self):
             s3plugin_lib.remove_if_exists(get_filename2)
             s3plugin_lib.remove_if_exists(get_filename3)
 
+    def multiple_irsync_to_s3__issue_2204(self):
```
Comment (alanking) [inline packaging/resource_suite_s3_nocache.py:2491]: Consider giving a more descriptive name to this test.

Also, is it an issue that this does not begin with the word `test`? I seem to recall `unittest` looking for functions that start with `test` or `test_` in order to consider it as a test... The same goes for `recursive_put_small_files_with_checksum_and_N_flag__issue_2198` above and the other new test below.

File: s3_resource/src/s3_operations.cpp (hunk 4 above)
Comment (JustinKyleJames) [reply s3_resource/src/s3_operations.cpp:336]: > Did we reach a conclusion about the `oprType` conditions?

No, unfortunately those oprType conditions were added while migrating away from S3FS.  It is just one commit now and I think I have lost the history of why they were added.  (Should have had a better comment.)

Tests are passing though and removing COPY_DEST did fix #2204.

File: packaging/resource_suite_s3_nocache.py (hunk 5 above)
Comment (JustinKyleJames) [reply packaging/resource_suite_s3_nocache.py]: Good catch.  The tests run when manually called but probably don't run when executing the entire suite.  I changed the names of the tests and ran them to make sure they still work.

**Review by alanking — COMMENTED** [review]
Please squash to taste and omit #s so we can take one more look and then I think this will be good to go.

**Comment by alanking** [comment]
Oh, and please apply clang-format if we're doing that over here.

**Comment by JustinKyleJames** [comment]
> Oh, and please apply clang-format if we're doing that over here.

I did.  I think all it did was insert tabs for spaces which I hate but anyway.

**Comment by JustinKyleJames** [comment]
> Please squash to taste and omit #s so we can take one more look and then I think this will be good to go.

done

**Review by alanking — APPROVED** [review]
Looks good to me. The other things clang-format is complaining about don't seem to have been touched by your changes, so, not sure what that's about.

Any other thoughts? @korydraughn or @trel ?

**Comment by JustinKyleJames** [comment]
> Looks good to me. The other things clang-format is complaining about don't seem to have been touched by your changes, so, not sure what that's about.
> 
> Any other thoughts? @korydraughn or @trel ?

Every time I run it it seems to expand the number of lines.  I did like four or five iterations and it kept doing that.  All the changes were spaces to tabs but it goes into things that were not touched.

I did push one more clang-format run before I realized what it was doing.

**Comment by korydraughn** [comment]
Ignore clang-format until the entire project is formatted. If you've already applied clang-format to your commits, that's fine.

@JustinKyleJames If the changes pass the tests, squash it. No pounds.

**Comment by JustinKyleJames** [comment]
> I did push one more clang-format run b

It is already fully squashed.  There are two commits to keep the two issues separate.

**Review by korydraughn — APPROVED** [review]
Cool cool. Let's add those pounds.

**Comment by JustinKyleJames** [comment]
> Cool cool. Let's add those pounds.

Done

**Pull request merged**